aif360/sklearn/detectors/facts/__init__.py

```diff
@@ -0,0 +1,418 @@
+            feats_not_allowed_to_change = set(all_feats) - set(self.feats_allowed_to_change)
+        elif self.feats_not_allowed_to_change is not None:
+            feats_not_allowed_to_change = set(self.feats_not_allowed_to_change)
+        else:
+            raise Exception("Code should be unreachable.")
+
+        num_features = list(set(X.columns) - set(self.categorical_features))
+        comparators = feature_change_builder(
+            X=X,
+            num_cols=num_features,
+            cate_cols=self.categorical_features,
+            ord_cols=[],
+            feature_weights=self.feature_weights,
+            num_normalization=False,
+        )
+        params = ParameterProxy(featureChanges=comparators)
+        
+        ifthens_coverage_correctness = valid_ifthens(
+            X=X,
+            model=self.clf,
+            sensitive_attribute=self.prot_attr,
+            freqitem_minsupp=self.freq_itemset_min_supp,
+            feats_not_allowed_to_change=list(feats_not_allowed_to_change),
+            verbose=verbose,
+        )
+
+        rules_by_if = rules2rulesbyif(ifthens_coverage_correctness)
+
+        if verbose:
+            print("Computing percentages of individuals flipped by any action with cost up to c, for every c", flush=True)
+        self.rules_with_cumulative = cum_corr_costs_all(
+            rulesbyif=rules_by_if,
+            X=X,
+            model=self.clf,
+            sensitive_attribute=self.prot_attr,
+            params=params,
+            verbose=verbose,
+        )
+        self.rules_by_if = calc_costs(rules_by_if)
+
+        self.dataset = X.copy(deep=True)
+
+        return self
+    
+    def bias_scan(
+        self,
+        metric: str = "equal-effectiveness",
+        viewpoint: str = "macro",
+        sort_strategy: str = "max-cost-diff-decr",
+        top_count: int = 10,
+        filter_sequence: List[str] = [],
+        phi: float = 0.5,
+        c: float = 0.5
+    ):
+        """Examines generated subgroups and calculates the `top_count` most
+        unfair ones, with respect to the chosen metric.
+
+        Stores the final groups in instance variable `self.top_rules` and the
+        respective subgroup costs in `self.subgroup_costs` (or `self.unfairness`
+        for the "fair-tradeoff" metric).
+
+        Args:
+            metric (str, optional): one of the following choices
+            
+                - "equal-effectiveness"
+                - "equal-choice-for-recourse"
+                - "equal-effectiveness-within-budget"
+                - "equal-cost-of-effectiveness"
+                - "equal-mean-recourse"
+                - "fair-tradeoff"
+                
+                Defaults to "equal-effectiveness".
+
+                For explanation of each of those metrics, refer either to the
+                paper [#FACTS23]_ or the demo_FACTS notebook.
+
+            viewpoint (str, optional): "macro" or "micro". Defaults to "macro".
```

Comment:
Add short explanation here.

Reply:
Done. I added the same short description which I also gave in a previous comment. Hopefully, it is sufficient, but I would be happy to make any extra modifications / clarifications you deem appropriate.